File: cpp/leetcodes/Count_Complete_Tree_Nodes.cpp
/*
Given a complete binary tree, count the number of nodes.

Definition of a complete binary tree from Wikipedia:
In a complete binary tree every level, except possibly the last,
 is completely filled, and all nodes in the last level are as far left as possible.
 It can have between 1 and 2h nodes inclusive at the last level h.
 */


#include <iostream>
#include <vector>
#include <bits/stl_algo.h>
#include <set>
#include <queue>
#include <stack>


using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int countNodes(TreeNode* root) {
        int height = heightOfTree(root);
        if (height <= 1){
            return height;
        }

        TreeNode* node = root;
        int ht = height;
        int count = pow(2, height-1)-1;
        TreeNode* parent = nullptr;
        while (node){
            if (rightHeightOfTree(node) == ht){
                count += pow(2, ht-1);
                break;
            }
            parent = node;
            node = node->left;
            ht--;
        }
        int h = 0;
        if (parent && parent->right){
            node = parent->right;
            while (node->left){
                node = node->left;
                h++;
            }
            if (h == ht-1){
                count++;
            }
        }

        return count;
    }

    int rightHeightOfTree(TreeNode* root){
        if(!root){
            return 0;
        }
        TreeNode* node = root;
        int h = 0;
        while (node){
            node = node->right;
            h++;
        }
        return h;
    }

    int heightOfTree(TreeNode* root){
        if(!root){
            return 0;
        }
        int h = 0;
        TreeNode* node = root;
        while (node){
            h++;
            node = node->left;
        }
        return h;
    }
};

int main(){
    Solution s;
    TreeNode* root = new TreeNode(1);
    cout << s.countNodes(root) << endl;

    root->left = new TreeNode(2);
    cout << s.countNodes(root) << endl;

    root->right = new TreeNode(3);
    cout << s.countNodes(root) << endl;

    root->left->left = new TreeNode(4);
    cout << s.countNodes(root) << endl;

    root->left->right = new TreeNode(5);
    cout << s.countNodes(root) << endl;

    root->right->left = new TreeNode(6);
    cout << s.countNodes(root) << endl;

    root->right->right = new TreeNode(7);
    cout << s.countNodes(root) << endl;

    root->left->left->left = new TreeNode(8);
    cout << s.countNodes(root) << endl;

    root = new TreeNode(1);
    queue<TreeNode*> q;
    q.push(root);
    int val = 2;
    while (val < 1000){
        queue<TreeNode*> nextLevel;
        while (!q.empty()){
            TreeNode* node = q.front();
            q.pop();
            node->left = new TreeNode(val++);
            cout << s.countNodes(root) << endl;

            node->right = new TreeNode(val++);
            cout << s.countNodes(root) << endl;

            nextLevel.push(node->left);
            nextLevel.push(node->right);
        }
        q = nextLevel;
    }

}

File: leetcode/medium/Count_Complete_Tree_Nodes.py
# -*- coding: utf-8 -*-
"""

Given a complete binary tree, count the number of nodes.

Definition of a complete binary tree from Wikipedia:
In a complete binary tree every level, except possibly the last,
is completely filled, and all nodes in the last level are as far left as possible.
It can have between 1 and 2h nodes inclusive at the last level h

"""
__author__ = 'huash'


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    """
    180ms
    Your runtime beats 94.83% of python coders.

    如果当前沿着最左边和沿着最右边走到底的高度是一样的，说明是一颗完全二叉树，结点数量是2^h-1;
    否则只可能左边高度大于右边，分别计算左右子树结点的数量。

    """
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        lh, rh = 1, 1
        lc, rc = root.left, root.right
        while lc:
            lh += 1
            lc = lc.left
        while rc:
            rh += 1
            rc = rc.right

        if lh == rh:
            return (1 << lh) - 1
        else:
            # lh > rh
            return 1 + self.countNodes(root.left) + self.countNodes(root.right)


    def countNodes2(self, root):
        """

        假设二叉树的高度是H，那么总结点数量是前H-1的结点数量+最后一层的结点数量。
        Sum = f(H-1) + g(H)
        f(h) = 2^h - 1

        问题转化为求最后一层结点的数量

        :param root:
        :return:
        """


    def brutal(self, root):
        """
        TLE
        :param root:
        :return:
        """
        if not root:
            return 0
        if not root.left and not root.right:
            return 1

        result = 1
        if root.left:
            result += self.brutal(root.left)
        if root.right:
            result += self.brutal(root.right)

        return result


s = Solution()
root = TreeNode(1)
print(s.brutal(root), s.countNodes(root))

root.left = TreeNode(2)
print(s.brutal(root), s.countNodes(root))

root.right = TreeNode(3)
root.left.left = TreeNode(4)
print(s.brutal(root), s.countNodes(root))

root.left.right = TreeNode(5)
root.right.left = TreeNode(6)
print(s.brutal(root), s.countNodes(root))

